EDDiscovery/UserControls/UserControlScanGrid.cs

```diff
@@ -0,0 +1,206 @@
+        }
+
+        private void Display(HistoryEntry he, HistoryList hl)            // Called at first start or hooked to change cursor
+        {
+            if (he != null && (last_he == null || he.System != last_he.System))
+            {
+                last_he = he;
+                DrawSystem();
+            }
+        }
+
+        void DrawSystem()   // draw last_sn, last_he
+        {
+            dataGridViewScangrid.Rows.Clear();
+
+            if (last_he == null)
+            {
+                SetControlText("No Scan");
+                return;
+            }
+
+
+            StarScan.SystemNode last_sn = discoveryform.history.starscan.FindSystem(last_he.System, true);
+
+            SetControlText((last_sn == null) ? "No Scan" : ("Brief Scan Summary for " + last_sn.system.name));
+
+            if (last_sn != null)
+            {
+                List<StarScan.ScanNode> all_nodes = new List<StarScan.ScanNode>();
+                foreach (StarScan.ScanNode starnode in last_sn.starnodes.Values)
+                {
+                    all_nodes = Flatten(starnode, all_nodes);
+                }
+
+
+                // flatten tree of scan nodes to prepare for listing
+                foreach (StarScan.ScanNode sn in all_nodes)
+                {
+
+                    
+                    // create the grid data
+
+                    // populate the body class
+                    StringBuilder bdClass = new StringBuilder();
+                    
+                    if (sn.ScanData != null && sn.ScanData.BodyName != null && sn.ScanData.PlanetClass != null)
+                    {
+                        bdClass.Append(sn.ScanData.PlanetClass);
+                    }
+                    if (sn.ScanData != null && sn.ScanData.BodyName != null && sn.ScanData.StarTypeText != null)
+                    {
+                        bdClass.Append(sn.ScanData.StarTypeText);
+                    }
+                    if (sn.level >= 2 && sn.type == StarScan.ScanNodeType.body)
+                    {
+                        bdClass.Append(" Moon");
+                    }
+
+                        // populate the detailed information
+                        StringBuilder bdDetails = new StringBuilder();
+
+                    if (sn.ScanData != null && sn.ScanData.BodyName != null && sn.ScanData.Terraformable == true)
+                    {
+                        bdDetails.Append("Terraformable. ");
+                    }
+                    if (sn.ScanData != null && sn.ScanData.BodyName != null && sn.ScanData.HasRings == true && sn.ScanData.IsStar == false)
+                    {
+                        bdDetails.Append("Ringed. ");
+                    }
+
+                    if (sn.ScanData != null && sn.ScanData.BodyName != null && sn.ScanData.IsLandable == true)
+                    {
+                        bdDetails.Append("Landable. ");
+                    }
+
+                    if (sn.ScanData.Volcanism != null)
```

Comment:
Either all of the `sn.ScanData != null` checks throughout are unwarranted, or this will throw a null reference exception if `sn.ScanData` were to be null. The same goes for the `.HasMaterials` checks immediately below.

Reply:
That was a commit that did not get merged, for some obscures reasons. My local branches has the right code here.

Reply:
Solved.